File: app/src/main/java/com/digicoffer/lauditor/Webservice/CommonApiHelper/MultiPartHttpExecute.java
package com.digicoffer.lauditor.Webservice.CommonApiHelper;

import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.util.Log;

import com.digicoffer.lauditor.LoginActivity.ViewModels.LoginActivity;
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener;
import com.digicoffer.lauditor.Webservice.HttpResultDo;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants;

import org.json.JSONObject;

import java.util.Iterator;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;


public class MultiPartHttpExecute extends AsyncTask<String, Integer, HttpResultDo> {

    public static AtomicInteger openCounter = new AtomicInteger(0);
    public static final int MAX_CONCURRENCY = 1;
    //    private HttpResultDo httpResult = null;
    private String requestId = null;
    private String requestType = null;
    private boolean sslFlag = false;
    WebServiceHelper.RestMethodType restMethodType = WebServiceHelper.RestMethodType.GET;
    private String URL = "";
    private AsyncTaskCompleteListener callback;
    private Context activity = null;

    HttpResultDo httpResult = new HttpResultDo();

    public MultiPartHttpExecute(String requestId, boolean sslFlag, WebServiceHelper.RestMethodType restMethodType, String baseURL,
                                AsyncTaskCompleteListener callback, Context activity, String requestType) {
        super();
        this.requestId = requestId;
        this.sslFlag = sslFlag;
        this.restMethodType = restMethodType;
        this.URL = baseURL;
        this.callback = callback;
        this.activity = activity;
        this.requestType = requestType;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
    }

    protected HttpResultDo doInBackground(String... params) {
        String data = "";
        String fullUrl = "";


        try {
            if (requestType.equals("Label") || (requestType.equals("auth")) || (requestType.equals("messages_rows")) || (URL.contains(Constants.EMAIL_UPLOAD_URL))) {
                fullUrl = URL;
            } else {
                fullUrl = Constants.base_URL + URL;
            }
            OkHttpClient client = new OkHttpClient().newBuilder()
                    .build();
            MediaType mediaType = MediaType.parse("application/json");
            RequestBody body = null;

            if (URL.equals("v3/decrypt")) {
                // Convert string to JSONObject
                JSONObject jsonObject = new JSONObject(params[0]);
                MultipartBody.Builder bodyBuilder = new MultipartBody.Builder();
                MultipartBody.Builder builder = bodyBuilder.setType(MultipartBody.FORM);

                for (Iterator<String> it = jsonObject.keys(); it.hasNext(); ) {
                    Object key = it.next();
                    String key_String = key.toString();
                    String key_value = jsonObject.getString(key_String);

                    builder.addFormDataPart(key_String, key_value);
                }
                body = builder.build();
            } else {
                body = RequestBody.create(mediaType, params[0]);
            }
            Request request = new Request.Builder()
                    .url(fullUrl)
                    .method(this.restMethodType.toString(), body)
                    .addHeader("Accept", "application/json")
                    .addHeader("Content-Type", "application/json")
                    .addHeader("Authorization", "Bearer " + Constants.TOKEN)
                    .build();
            // Perform the request asynchronously
            Response response = client.newCall(request).execute();

            int status_code = response.code();
            Log.i("status_code:", String.valueOf(status_code));
            httpResult.setStatus_code(status_code);
            // Check if the response is successful
            if (response.isSuccessful()) {
                // Handle successful response
                assert response.body() != null;
                String responseData = response.body().string();
                Log.d("Response", responseData);
                httpResult.setResult(WebServiceHelper.ServiceCallStatus.Success);
                httpResult.setResponseContent(responseData);
                return httpResult;
            } else {
                // Handle unsuccessful response
                assert response.body() != null;
                String responseData = response.body().string();
                Log.e("Response", "Error: " + response.code());
                Log.e("Response", "Body: " + responseData);
                httpResult.setResult(WebServiceHelper.ServiceCallStatus.Failed);
                httpResult.setResponseContent(responseData);
                return httpResult;
            }

        } catch (Exception e) {
            AndroidUtils.logMsg("HttpExecuteTask.doInBackground(): Exception " + e.getMessage());
            httpResult = new HttpResultDo();
            httpResult.setResult(WebServiceHelper.ServiceCallStatus.Exception);
            httpResult.setResponseContent(e.getMessage());
            return httpResult;
        } finally {

        }
    }

    protected void onPostExecute(HttpResultDo httpResult) {
        openCounter.decrementAndGet();
        httpResult.setRequestId(requestId);
        httpResult.setRequestType(requestType);
//        AndroidUtils.logMsg("HttpExecuteTask.onPostExecute() " + httpResult);
        super.onPostExecute(httpResult);
        try {
            Log.e("Response_Msg", httpResult.getResponseContent());
            if (httpResult.getStatus_code() == 401 && !(requestType.equals("Label")) && !(requestType.equals("auth"))
                    && !(requestType.equals("messages_rows")) && !(URL.contains(Constants.EMAIL_UPLOAD_URL)) && !(requestType.equals("Dashboard"))) {
                Intent in = new Intent(activity, LoginActivity.class);
                in.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                activity.startActivity(in);
                AndroidUtils.showToast("Session expired, Login again", activity);
            } else if ((requestType.equals("Dashboard")) && (httpResult.getStatus_code() == 401)) {
                Constants.Valid_Token = false;
//                if (!token.isEmpty()) {
//                    Constants.loginActivity.Dashboard();
//                } else if (!email.isEmpty() && !password.isEmpty()) {
//                if (!Constants.Email.isEmpty() && !Constants.OLD_PASSWORD.isEmpty()) {
//                    Constants.loginActivity.Login(Constants.Email, Constants.OLD_PASSWORD);
//                } else {
                    Log.d("Session", "Session expired, Login again");
                    Intent nav_bio = new Intent(activity, LoginActivity.class);
                    activity.startActivity(nav_bio);
//                    AndroidUtils.showToast("Session expired, Login again", activity);
//                    Intent nav_bio = new Intent(activity, LoginActivity.class);
//                    activity.startActivity(nav_bio);
//                }
//                }
//                Intent nav_bio;
//                if (Constants.Biometric_checked) {
//                    nav_bio = new Intent(activity, biometric_page.class);
//                } else {
//                    nav_bio = new Intent(activity, LoginActivity.class);
//                }
//                activity.startActivity(nav_bio);
            } else
                Constants.Valid_Token = true;
            callback.onAsyncTaskComplete(httpResult);
        } catch (Exception e) {
            Log.d("Error_mg", Objects.requireNonNull(e.getMessage()));
            AndroidUtils.logMsg("HttpExecuteTask.onPostExecute() : Exception " + e.getMessage());
        }
    }
}

File: app/src/main/java/com/digicoffer/lauditor/Webservice/CommonApiHelper/MultiPartHttpExecute.kt
package com.digicoffer.lauditor.Webservice.CommonApiHelper

import android.content.Context
import android.content.Intent
import android.os.AsyncTask
import android.util.Log
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants
import com.digicoffer.lauditor.LoginActivity.ViewModels.LoginActivity
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener
import com.digicoffer.lauditor.Webservice.HttpResultDo
import com.digicoffer.lauditor.Webservice.core.RetrofitClient
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody
import okhttp3.ResponseBody
import org.json.JSONObject
import retrofit2.Call
import retrofit2.Response
import java.io.IOException
import java.util.HashMap
import java.util.concurrent.atomic.AtomicInteger

class MultiPartHttpExecute(
    private val requestId: String,
    private val sslFlag: Boolean,
    private val restMethodType: WebServiceHelper.RestMethodType,
    private val URL: String,
    private val callback: AsyncTaskCompleteListener,
    private val activity: Context,
    private val requestType: String
) : AsyncTask<String, Int, HttpResultDo>() {

    companion object {
        @JvmField
        val openCounter = AtomicInteger(0)
        const val MAX_CONCURRENCY = 1
    }

    private val httpResult = HttpResultDo()

    override fun onPreExecute() {
        super.onPreExecute()
    }

    override fun doInBackground(vararg params: String?): HttpResultDo {
        var fullUrl = ""
        try {
            fullUrl = if (requestType == "Label" || requestType == "auth" || requestType == "messages_rows" || URL.contains(Constants.EMAIL_UPLOAD_URL)) {
                URL
            } else {
                Constants.base_URL + URL
            }

            val bodyParam = if (params.isNotEmpty() && params[0] != null) params[0]!! else ""

            ApiMonitorLogger.logSending(
                requestType = requestType,
                url = fullUrl,
                method = restMethodType.name,
                params = bodyParam
            )

            val headers = HashMap<String, String>()
            headers["Accept"] = "application/json"
            headers["Authorization"] = "Bearer ${Constants.TOKEN}"

            var requestBody: RequestBody? = null
            if (URL == "v3/decrypt") {
                val jsonObject = JSONObject(bodyParam)
                val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
                val keys = jsonObject.keys()
                while (keys.hasNext()) {
                    val key = keys.next()
                    val value = jsonObject.getString(key)
                    builder.addFormDataPart(key, value)
                }
                requestBody = builder.build()
            } else {
                headers["Content-Type"] = "application/json"
                val mediaType = "application/json".toMediaTypeOrNull()
                requestBody = RequestBody.create(mediaType, bodyParam)
            }

            val apiService = RetrofitClient.getApiService()
            val call: Call<ResponseBody> = when (restMethodType) {
                WebServiceHelper.RestMethodType.GET -> apiService.executeGet(fullUrl, headers)
                WebServiceHelper.RestMethodType.POST -> apiService.executePost(fullUrl, headers, requestBody)
                WebServiceHelper.RestMethodType.PUT -> apiService.executePut(fullUrl, headers, requestBody)
                WebServiceHelper.RestMethodType.PATCH -> apiService.executePatch(fullUrl, headers, requestBody)
                WebServiceHelper.RestMethodType.DELETE -> {
                    if (bodyParam.isNotEmpty()) {
                        apiService.executeDeleteWithBody(fullUrl, headers, requestBody)
                    } else {
                        apiService.executeDelete(fullUrl, headers)
                    }
                }
            }

            val response = call.execute()
            val statusCode = response.code()
            Log.i("status_code:", statusCode.toString())
            httpResult.status_code = statusCode

            val responseData = if (response.isSuccessful) {
                httpResult.result = WebServiceHelper.ServiceCallStatus.Success
                val bodyString = response.body()?.string() ?: ""
                Log.d("Response", bodyString)
                bodyString
            } else {
                httpResult.result = WebServiceHelper.ServiceCallStatus.Failed
                val errorString = response.errorBody()?.string() ?: ""
                Log.e("Response", "Error: $statusCode")
                Log.e("Response", "Body: $errorString")
                errorString
            }
            httpResult.responseContent = responseData

        } catch (e: Exception) {
            AndroidUtils.logMsg("MultiPartHttpExecute.doInBackground(): Exception ${e.message}")
            httpResult.result = WebServiceHelper.ServiceCallStatus.Exception
            httpResult.responseContent = e.message ?: "Unknown error"
        }
        return httpResult
    }

    override fun onPostExecute(result: HttpResultDo) {
        openCounter.decrementAndGet()
        result.requestId = requestId
        result.requestType = requestType
        super.onPostExecute(result)

        try {
            Log.e("Response_Msg", result.responseContent)
            val fullUrl = if (requestType == "Label" || requestType == "auth" || requestType == "messages_rows" || URL.contains(Constants.EMAIL_UPLOAD_URL)) {
                URL
            } else {
                Constants.base_URL + URL
            }

            ApiMonitorLogger.logResponse(
                requestType = requestType,
                url = fullUrl,
                params = null,
                httpResult = result
            )

            val statusCode = result.status_code

            if (statusCode == 401 && requestType != "Label" && requestType != "auth" &&
                requestType != "messages_rows" && !URL.contains(Constants.EMAIL_UPLOAD_URL) && requestType != "Dashboard") {
                val intent = Intent(activity, LoginActivity::class.java).apply {
                    addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_NEW_TASK)
                }
                activity.startActivity(intent)
                AndroidUtils.showToast("Session expired, Login again", activity)
            } else if (requestType == "Dashboard" && statusCode == 401) {
                Constants.Valid_Token = false
                Log.d("Session", "Session expired, Login again")
                val intent = Intent(activity, LoginActivity::class.java)
                activity.startActivity(intent)
            } else {
                Constants.Valid_Token = true
            }

            callback.onAsyncTaskComplete(result)
        } catch (e: Exception) {
            Log.d("Error_mg", e.message ?: "Null message")
            AndroidUtils.logMsg("MultiPartHttpExecute.onPostExecute() : Exception ${e.message}")
        }
    }
}
